runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -236,6 +262,96 @@ function M.on_publish_diagnostics(_, result, ctx, config)
   vim.diagnostic.set(namespace, bufnr, diagnostic_lsp_to_vim(diagnostics, bufnr, client_id))
 end
 
+---@private
+local function resolve_bufnr(bufnr)
+  return bufnr > 0 and bufnr or api.nvim_get_current_buf()
+end
+
+--- Refresh diagnostics for a buffer
+---
+---@param opts (nil|table) Optional arguments
+---  - bufnr (integer, default: 0): Buffer whose hints to refresh
+---  - only_visible (boolean, default: false): Whether to only refresh hints for the visible regions of the buffer
+---
+---@private
+function M.refresh(opts)
+  opts = opts or {}
+  local bufnr = resolve_bufnr(opts.bufnr or 0)
+  local bufstate = bufstates[bufnr]
+  if not (bufstate and bufstate.enabled) then
+    return
+  end
+  local only_visible = opts.only_visible or false
+  local buffer_windows = {}
+  for _, winid in ipairs(api.nvim_list_wins()) do
+    if api.nvim_win_get_buf(winid) == bufnr then
+      table.insert(buffer_windows, winid)
+    end
+  end
+  for _, window in ipairs(buffer_windows) do
+    local first = vim.fn.line('w0', window)
+    local last = vim.fn.line('w$', window)
+    local params = {
+      textDocument = util.make_text_document_params(bufnr),
+      range = {
+        start = { line = first - 1, character = 0 },
+        ['end'] = { line = last, character = 0 },
+      },
+    }
+    vim.lsp.buf_request(bufnr, 'textDocument/diagnostic', params)
+  end
+  if not only_visible then
+    local params = {
+      textDocument = util.make_text_document_params(bufnr),
+      range = {
+        start = { line = 0, character = 0 },
+        ['end'] = { line = api.nvim_buf_line_count(bufnr), character = 0 },
+      },
+    }
+    vim.lsp.buf_request(bufnr, 'textDocument/diagnostic', params)
+  end
+end
+
+function M.on_diagnostic(_, result, ctx, config)
+  local client_id = ctx.client_id
+  local uri = ctx.params.textDocument.uri
+  local fname = vim.uri_to_fname(uri)
+  if result.kind == "unchanged" then
+    return
+  end
+
+  local diagnostics = result.items
+  if #diagnostics == 0 and vim.fn.bufexists(fname) == 0 then
+    return
+  end
+  local bufnr = vim.fn.bufadd(fname)
+
+  if not bufnr then
+    return
+  end
+
+  client_id = get_client_id(client_id)
+
+  local namespace = M.get_namespace(client_id)
```

Comment:
I think we'll need to use the `identifier` from [DiagnosticOption](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnosticOptions) as an additional distinction.

See also https://github.com/microsoft/language-server-protocol/issues/1743
Otherwise this may conflict with the pushed diagnostics